pkg/database/handler.go

```diff
@@ -221,6 +228,123 @@ func (h Handler) SaveAs(c *gin.Context) {
 	c.JSON(http.StatusCreated, save)
 }
 
+// Save database
+func (h Handler) Save(c *gin.Context) {
+	// swagger:route POST /databases/save/{instanceId} saveDatabase
+	//
+	// Save database
+	//
+	// Saving a database involves the following steps
+	//
+	// 1. Lock database if not already locked
+	// 2. Save as
+	// 3. Delete the existing database
+	// 4. Rename the saved database in the file storage
+	// 5. Overwrite the saved as database's properties with the ones from the old database (including the id)
+	// 6. Unlock the database if it wasn't already locked
+	//
+	// This won't affect the instance running the database. However, it should be noted that if two databases are deployed from the same database they can both overwrite it. It's up to the users to ensure this doesn't happen accidentally.
+	//
+	// Security:
+	//	oauth2:
+	//
+	// Responses:
+	//	202:
+	//	401: Error
+	//	403: Error
+	//	404: Error
+	//	415: Error
+	instanceIdParam := c.Param("instanceId")
+	instanceId, err := strconv.ParseUint(instanceIdParam, 10, 32)
+	if err != nil {
+		badRequest := apperror.NewBadRequest("error parsing instanceId")
+		_ = c.Error(badRequest)
+		return
+	}
+
+	instance, err := h.instanceService.FindByIdDecrypted(uint(instanceId))
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	stack, err := h.stackService.Find(instance.StackName)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	databaseIdString, err := findParameter("DATABASE_ID", instance, stack)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	databaseId, err := strconv.ParseUint(databaseIdString, 10, 32)
+	if err != nil {
+		badRequest := apperror.NewBadRequest("error parsing databaseId")
+		_ = c.Error(badRequest)
+		return
+	}
+
+	database, err := h.databaseService.FindById(uint(databaseId))
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	err = h.canAccess(c, database)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	user, err := handler.GetUserFromContext(c)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	lock := database.Lock
+	isLocked := lock != nil
+	if isLocked && (lock.InstanceID != uint(instanceId) || lock.UserID != user.ID) {
+		unauthorized := apperror.NewUnauthorized("database is locked")
+		_ = c.Error(unauthorized)
+		return
+	}
+
+	if !isLocked {
+		_, err := h.databaseService.Lock(uint(databaseId), uint(instanceId), user.ID)
+		if err != nil {
+			_ = c.Error(err)
+			return
+		}
+	}
+
+	err = h.databaseService.Save(database, instance, stack)
```

Comment:
It looks to me as if we are potentially leaving the DB in a locked state if any of the operations from here on return an error. We should probably unlock in a defer.